pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+		#[frame_support::transactional]
+		pub fn push_withdraw_in_queue(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			nft: &mut NftAttr<BalanceOf<T>>,
+			nft_id: NftId,
+			userid: T::AccountId,
+			shares: BalanceOf<T>,
+			maybe_vault_pid: Option<u64>,
+		) -> DispatchResult {
+			// Remove the existing withdraw request in the queue if there is any.
+			let (in_queue_nfts, new_withdraw_queue): (VecDeque<_>, VecDeque<_>) = base_pool
+				.withdraw_queue
+				.clone()
+				.into_iter()
+				.partition(|withdraw| withdraw.user == userid);
+			// only one nft withdraw request should be in the queue
+			base_pool.withdraw_queue = new_withdraw_queue;
+			let price = base_pool
+				.share_price()
+				.expect("In withdraw case, price should always exists;");
+
+			match maybe_vault_pid {
+				Some(pid) => {
+					let mut vault = ensure_vault::<T>(pid)?;
+					for withdrawinfo in &in_queue_nfts {
+						let in_queue_nft = Self::get_nft_attr(base_pool, withdrawinfo.nft_id)
+							.expect("get nft attr should always success; qed.");
+						let withdraw_amount = bmul(in_queue_nft.shares.clone(), &price);
+						nft.stakes += in_queue_nft.stakes.min(withdraw_amount);
+						nft.shares += in_queue_nft.shares;
+						Self::burn_nft(base_pool, withdrawinfo.nft_id)
+							.expect("burn nft attr should always success; qed.");
+						vault.basepool.total_stake =
+							vault.basepool.total_stake - in_queue_nft.stakes + withdraw_amount;
+						vault.basepool.free_stake += withdraw_amount;
+					}
+					PoolCollection::<T>::insert(
+						pid,
+						PoolProxy::<T::AccountId, BalanceOf<T>>::Vault(vault.clone()),
+					);
+				}
+				None => {
+					for withdrawinfo in &in_queue_nfts {
+						let in_queue_nft = Self::get_nft_attr(base_pool, withdrawinfo.nft_id)
+							.expect("get nft attr should always success; qed.");
+						let withdraw_amount = bmul(in_queue_nft.shares.clone(), &price);
+						nft.stakes += in_queue_nft.stakes.min(withdraw_amount);
+						nft.shares += in_queue_nft.shares;
+						Self::burn_nft(base_pool, withdrawinfo.nft_id)
+							.expect("burn nft attr should always success; qed.");
+					}
+				}
+			};
+
+			let amount = bmul(shares, &price);
+			let split_nft_id = Self::mint_nft(base_pool, pallet_id(), shares, amount)
+				.expect("mint nft should always success");
+			nft.shares = nft
+				.shares
+				.checked_sub(&shares)
+				.ok_or(Error::<T>::InvalidShareToWithdraw)?;
+			nft.stakes = nft
+				.stakes
+				.checked_sub(&amount)
+				.ok_or(Error::<T>::InvalidShareToWithdraw)?;
+			// Push the request
+			let now = <T as registry::Config>::UnixTime::now()
+				.as_secs()
+				.saturated_into::<u64>();
+			base_pool.withdraw_queue.push_back(WithdrawInfo {
+				user: userid.clone(),
+				start_time: now,
+				nft_id: split_nft_id,
+			});
+
+			Ok(())
+		}
+
+		pub fn has_expired_withdrawal(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
```

Comment:
`base_pool` is read only. Use immutable borrow instead

Reply:
done